### PR TITLE
MAINT: Fix expm1 instability for small complex numbers.

### LINKED ISSUE #14019: Numerical instability for expm1 with complex arguments
The output of `np.expm1` becomes unstable for small values (around 10^{-15} to 10^{-16}) when the argument is complex. This does not seem to be an issue with the numerical precision associated with the real and imaginary components: using a `float32` data type works as expected. 

The problem arises because of the way `expm1` is [currently implemented](https://github.com/numpy/numpy/blob/deea4983aedfa96905bbaee64e3d1de84144303f/numpy/core/src/umath/funcs.inc.src#L360-L367) for complex arguments. In particular:

```c
static void
nc_expm1@c@(@ctype@ *x, @ctype@ *r)
{
    @ftype@ a = npy_exp@c@(x->real);
    // If x->imag == 0, then we evaluate exp(x->real) - 1 naively 
    // rather than using a more stable implementation
    r->real = a*npy_cos@c@(x->imag) - 1.0@c@;
    r->imag = a*npy_sin@c@(x->imag);
    return;
}
```

### Reproducing code example:

```python
import numpy as np
from matplotlib import pyplot as plt

fig, ax = plt.subplots()

x = np.logspace(-16.2, -14)
xs = [
    x,
    x.astype(np.complex),
    x.astype(np.float32),
]

ys = [np.expm1(x_) for x_ in xs]

for x, y in zip(xs, ys):
    ax.plot(x, y, marker='.', label=f'{x.dtype}: np.expm(x)')

ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlabel('x')
ax.legend(loc='upper left')

fig.tight_layout()
```

Virtually identical behaviour can be observed by replacing `ys = [np.expm1(x_) for x_ in xs]` with `ys = [-np.expm1(-x_) for x_ in xs]`, i.e. the problem is not unique to positive values of the argument.

image

### Error message:

No error message to report. However, the expected behaviour would be for the complex datatype to behave in the same fashion as the float32 and float64 data types.

### Numpy/Python version information:

```
1.16.4 3.7.2 (default, Jul 11 2019, 09:18:04) 
[Clang 10.0.1 (clang-1001.0.46.4)]
```

### PR DESCRIPTION
**Pull request by tillahoffmann**
Fixes #14019 based on the [scipy evaluation](https://github.com/scipy/scipy/blob/557ab7e/scipy/special/_cunity.pxd#L86-L115) by reordering terms.

### CODE REVIEW

File: numpy/core/src/umath/funcs.inc.src [hunk 1]
```diff
@@ -360,9 +360,9 @@ nc_exp2@c@(@ctype@ *x, @ctype@ *r)
 static void
 nc_expm1@c@(@ctype@ *x, @ctype@ *r)
 {
-    @ftype@ a = npy_exp@c@(x->real);
-    r->real = a*npy_cos@c@(x->imag) - 1.0@c@;
-    r->imag = a*npy_sin@c@(x->imag);
+    @ftype@ a = npy_cos@c@(x->imag);
+    r->real = npy_expm1@c@(x->real) * a + (a - 1);
```
Comment (charris) [inline numpy/core/src/umath/funcs.inc.src]: I wonder if `cos(imag) - 1` is sufficiently accurate? `-2 * sin(imag/2)**2` would have better precision. `cexpm1` is also an optional c99 function, we might want to track its status long term.

Comment (charris) [reply numpy/core/src/umath/funcs.inc.src]: I did find [this](https://tinyurl.com/v4au7aq). Maybe not worth worrying about right now, but this function could probably use further improvement.

Comment (tillahoffmann) [reply numpy/core/src/umath/funcs.inc.src]: Good point regarding using the trig identities to improve the numerics. Will update that in the code. Are you happy to leave the insights from the link you sent in a future PR?

image

Comment (charris) [reply numpy/core/src/umath/funcs.inc.src]: > Are you happy to leave the insights from the link you sent to a future PR?

Yes. I think this is already a significant improvement.

**Comment by charris** [comment]
If you do implement the more involved algorithm the test could be made more complete by using a larger set of small numbers for both the real and imaginary parts of the argument. Now that we are requiring C99, there are probably other cleanups that could be made to our complex function implementations.

**Comment by charris** [comment]
Thanks @tillahoffmann .

**Pull request merged**